pkg/app/gnmi_server.go

```diff
@@ -682,41 +690,50 @@ func (a *App) handleStreamSubscriptionRequest(sc *streamClient) {
 					SuppressRedundant: sub.GetSuppressRedundant(),
 					UpdatesOnly:       sc.req.GetSubscribe().GetUpdatesOnly(),
 				}
-				a.Logger.Printf("cache subscribe: %+v", ro)
-				for n := range a.c.Subscribe(sc.stream.Context(), ro) {
-					if n.Err != nil {
-						errChan <- n.Err
-						a.Logger.Printf("cache subscribe failed: %+v: %v", ro, n.Err)
-						return
-					}
-					err := sc.stream.Send(&gnmi.SubscribeResponse{
-						Response: &gnmi.SubscribeResponse_Update{
-							Update: n.Notification,
-						},
-					})
-					if err != nil {
-						errChan <- n.Err
-						return
-					}
+			}
+
+			a.Logger.Printf("cache subscribe: %+v", ro)
+
+			for n := range a.c.Subscribe(ctx, ro) {
+				// `errChan <- n.Err` should trigger the gnmi-server side cleanup
+				// only wait would be for the cache to close the channel
+				if n.Err != nil {
+					errChan <- n.Err
+					a.Logger.Printf("cache subscribe failed: %+v: %v", ro, n.Err)
+
+					// reader should only stop once the channel is closed by sender or otherwise
+					// it coould block the senders who doesn't know that error has happened
+
+					continue
+				}
+
+				err := sc.stream.Send(&gnmi.SubscribeResponse{
+					Response: &gnmi.SubscribeResponse_Update{
+						Update: n.Notification,
+					},
+				})
+
+				if err != nil {
+					errChan <- n.Err
 				}
-				return
 			}
 		}(sub)
 	}
 
 	// wait for ctx to be done
-	<-sc.stream.Context().Done()
-	errChan <- sc.stream.Context().Err()
+	<-ctx.Done()
+	errChan <- ctx.Err()
 	wg.Wait()
 }
 
 func (a *App) handlePolledSubscription(sc *streamClient) {
-	a.handleONCESubscriptionRequest(sc)
 	defer close(sc.errChan)
+	a.handleONCESubscriptionRequest(sc)
```

Comment:
any reason to switch the `defer` and `handleONCESubscriptionRequest` calls sequence?

Reply:
typo again